crates/ra_syntax/src/ast/extensions.rs

```diff
@@ -37,6 +35,41 @@ fn text_of_first_token(node: &SyntaxNode) -> &SmolStr {
     node.green().children().first().and_then(|it| it.as_token()).unwrap().text()
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CfgAttrKind {
+    Invalid,
+    Atom(SmolStr),
+    KeyValue { key: SmolStr, value: SmolStr },
+    All(Vec<CfgAttrKind>),
+    Any(Vec<CfgAttrKind>),
+    Not(Box<CfgAttrKind>),
+}
```

Comment:
This belongs to `hir` I think: there's nothing *syntactically* special about `cfg` attributes, they are just token trees.

Reply:
And looking at the existing helpers, maybe we want to move some of them to hir as well? Unsure. However, cfg grammar is definitely something to specific for syntax

Reply:
Hm, so, let me try to better explain the design I have in mind. 

First, we need CfgOptions and Expressions. I think it would be simplest to put them into `ra_db`, although they could live in a separate crate even (I wonder if we can share this hypothetical cfg crate between rustc, cargo and rust-analyzer...).

Then, as we don't want to interpret attributes directly in raw items (b/c attrs can be macros, and to interpret them we need more info than there's in raw items), we should lower syntactic attrs into token streams at the raw items layer. 

Finally, at the `DefCollector` layer, we parse attrs token streams into `CfgExpr`s and process them.

Does this all sound reasonable?

Reply:
I'm not sure where to place `CfgExpr`. But it **is** a [special builtin syntax](https://doc.rust-lang.org/reference/conditional-compilation.html#conditional-compilation), which is used in `#[cfg()]`, `#[cfg_attr()]`, `cfg!()` and the `target` selector in `Cargo.toml` like `[target.'cfg(not(windows))'.dependencies]`.

You're right, handling in `DefCollector` layer is better. And it seems not easy to get attribute AST in `DefCollector` through [`raw::RawItem`](https://github.com/rust-analyzer/rust-analyzer/blob/df954f2c581c6386390981b29cda8a33b92b0a5c/crates/ra_hir/src/nameres/raw.rs#L129). `ImportData` and `DefData` contain no AST information. Should we store  attribute (or the AST of item?) in both of them, and extract them from `raw::RawItem`?

Reply:
I would say it’s a build-in *feature*, but not syntax, as it just re-uses general attribute syntax. It’s by desig that it’s hard to get from DefCollector to the AST. This helps with incrementalism, as AST changes with every file modifications, while RawItems stay mostly the same. So we need to thread attribute info explicitly via raw items. I think he best way to do it is to do the same thing we do for macros: by turning an AST into a token tree and storing it as a part of raw item.

Reply:
Okay. 

> we need CfgOptions and Expressions. I think it would be simplest to put them into ra_db

Is it quite weird? If we store `CfgExpr` in db crate.

Reply:
It is indeed weird! Perhaps it’s better to move this to a separate crate from the start? ra_cfg will have CfgExpr, CfgOptoins, and a parser from tt::TokenStream to CfgExpr. That is, it will only depend on ra_tt. ra_db will then depend on ra_cfg for CfgOptions.